### ISSUE TITLE
dplyr::distinct requires .keep-all=TRUE to preserve its behavior

### ISSUE THREAD

**Issue by iDemirsoy** [dhimmel/SIDER2#1]
Hi there,
I was trying to get on side effect similarities between pairs of drug but similarity.df gives me this error. 

```
Error: Can't rename columns that don't exist. x Column '`similarity`' doesn't exist.
```

I was trying to dig into and i realized that `GetSimilarityDF` only returns 2 dimensional vector compound_1 and compound_2 therefore, there is no 3rd column to name as `'similarity'`. 

Therefore, I fixed the code as 

```r
GetSimilarityDF <- function(mat, w) {
  mat %>%
    proxy::simil(
      method = SiderSimilarity, w = w,
      by_rows = TRUE, upper = FALSE, diag = FALSE) %>%
    as.matrix() %>%
    reshape2::melt(na.rm = TRUE, value.name = 'similarity', varnames = c('compound_1', 'compound_2')) %>%
    dplyr::filter(compound_1 != compound_2) %>% # remove self-similarity (for safety, since `diag = FALSE` in proxy::simil)
    dplyr::distinct(compound_1, compound_2 ,similarity) # proxy::simil returning a non-NA upper half, so duplicate rows exist
}
```

Thanks for the sharing the code,

**Comment by dhimmel** [comment]
Thanks for the report and solution!

Here is the source code in question:

https://github.com/dhimmel/SIDER2/blob/9d585685dbeaba3bbac58024c814ac87521122ad/similarity/index.Rmd#L216-L225

Looks like dplyr 0.5.0 from 2016-06-24 changed the behavior of `dplyr::distinct`:

> [`distinct()`](https://dplyr.tidyverse.org/reference/distinct.html) now only keeps the distinct variables. If you want to return all variables (using the first row for non-distinct values) use `.keep_all = TRUE` ([#1110](https://github.com/tidyverse/dplyr/issues/1110)).

I think the best solution might be to add `.keep_all = TRUE`, which should replicate the pre-0.5.0 behavior.

Do you want to submit a pull request that adds this argument?

**Comment by iDemirsoy** [comment]
No, I would just fix it in my end. I was wondering if your sider4 can compute side effect similarity? I looked at the code but I did not see anything to attempt for computing. Just wondering. Maybe I should email for this, do you plan to compute other similarities  between pairs of drug such as protein. enzyme so on ?

**Comment by dhimmel** [comment]
> I was wondering if your sider4 can compute side effect similarity?

It doesn't look like https://github.com/dhimmel/SIDER4 computes side effect similarity.

> Maybe I should email for this

I prefer GitHub Issues.

> do you plan to compute other similarities between pairs of drug such as protein. enzyme so on ?

No. I am not currently working on or maintaining the SIDER2 and SIDER4 projects. My most recent work in the area is described at https://greenelab.github.io/connectivity-search-manuscript/ (still a work in progress). But this research does not directly compute similarity based on side effects.

Still happy to answer questions and receive bug reports like this on existing code.
